rasa_core/policies/ensemble.py

```diff
@@ -10,84 +10,89 @@
     from rasa_core.domain import Domain
     from rasa_core.featurizers import Featurizer
+    from rasa_core.policies.policy import Policy
+    from rasa_core.trackers import DialogueStateTracker
 
 
 class PolicyEnsemble(object):
     def __init__(self, policies, action_fingerprints=None):
+        # type: (List[Policy], Optional[Dict]) -> None
         self.policies = policies
-        self.training_metadata = {}
+        self.training_events = None
 
         if action_fingerprints:
             self.action_fingerprints = action_fingerprints
         else:
             self.action_fingerprints = {}
 
-    def max_history(self):
-        # type: () -> Optional[int]
-        """Return max history, only works if the ensemble is already trained."""
-
-        if self.policies:
-            return self.policies[0].max_history
-        else:
-            return None
-
-    def train(self, training_data, domain, featurizer, **kwargs):
-        # type: (DialogueTrainingData, Domain, Featurizer, **Any) -> None
-        if not training_data.is_empty():
+    def train(self, training_trackers, training_events, domain, **kwargs):
+        # type: (List[DialogueStateTracker], Dict, Domain, **Any) -> None
+        if training_trackers:
             for policy in self.policies:
-                policy.prepare(featurizer,
-                               max_history=training_data.max_history())
-                policy.train(training_data, domain, **kwargs)
-                self.training_metadata.update(training_data.metadata)
+                policy.train(training_trackers, domain, **kwargs)
+            self.training_events = training_events
         else:
             logger.info("Skipped training, because there are no "
                         "training samples.")
 
+    def probabilities_using_best_policy(self, tracker, domain):
+        # type: (DialogueStateTracker, Domain) -> List[float]
+        raise NotImplementedError
+
     def predict_next_action(self, tracker, domain):
-        # type: (DialogueStateTracker, Domain) -> (float, int)
+        # type: (DialogueStateTracker, Domain) -> int
         """Predicts the next action the bot should take after seeing x.
 
         This should be overwritten by more advanced policies to use ML to
         predict the action. Returns the index of the next action"""
         probabilities = self.probabilities_using_best_policy(tracker, domain)
-        max_index = np.argmax(probabilities)
+        max_index = int(np.argmax(probabilities))
         logger.debug("Predicted next action '{}' with prob {:.2f}.".format(
                 domain.action_for_index(max_index).name(),
                 probabilities[max_index]))
         return max_index
 
-    def probabilities_using_best_policy(self, tracker, domain):
-        raise NotImplementedError
+    def _max_histories(self):
+        # type: () -> List[Optional[int]]
+        """Return max history, only works if the ensemble is already trained."""
```

Comment:
can we raise an exception if it is not?

Reply:
if it is not = if the ensemble is trained?

Reply:
that's actually wrong description, it would work before training as well